### PR TITLE
PolyhedronROI based on gmsh API

### PR DESCRIPTION
**Pull request by castigli**
Added an implementation of PolyhedronROI which tags directly gmsh meshes (*.msh) and does not depends on STEPS.

### CODE REVIEW

File: polyhedronROI/__init__.py [hunk 1]
```diff
@@ -1,3 +1 @@
-from .annotation import gen_tet_spatial_index, add_tet_ROIs
-
 __version__ = "1.0.0"
```
Comment (WeiliangChenOIST) [inline polyhedronROI/__init__.py:1]: update the version?

**Comment by WeiliangChenOIST** [comment]
two major issues:
* if the mesh was tagged before, the old tags are removed after processing.
* If we tag the tets with 2 ROIs, the number of tets seems to be doubled in the output mesh, and gmsh reports many duplicated elements

**Comment by castigli** [comment]
> two major issues:
> 
>     * if the mesh was tagged before, the old tags are removed after processing.
> 
>     * If we tag the tets with 2 ROIs, the number of tets seems to be doubled in the output mesh, and gmsh reports many duplicated elements

I addressed the first point, the old physical entities are now retained (only 3D entities).

Instead the duplicate tets issue is a more fundamental problem the we need to discuss how to resolve.
 What you call a tag  is called physical group in gmsh,  a physical group can be made of one or more entities (essentially groups of tets in our case). As far as I understand a tet needs to be duplicated to belong to more than one entity.
In the case in which there are 2 overlapping ROIs which, in practice it means that there are 3 different volumes  with 3 distinct physical behaviors (let's say phys_a, phys_b, phys_c=phys_a+phys_b).
I think we might need to revisit the tagging strategy/meaning.

**Comment by WeiliangChenOIST** [comment]
yeah, that is what I've worried about. In the legacy 2.0 format it is possible to assign an element with multiple tags http://gmsh.info/doc/texinfo/gmsh.html#MSH-file-format-version-2-_0028Legacy_0029 
```
$Elements
number-of-elements
elm-number elm-type number-of-tags < tag > … node-number-list
…
$EndElements
```
but from what I get the new format drop that supports and opt for single tag strategy. In practice, as compartments/patches  are physical identities,  single-tag solution should be fine. But for ROI, as an element can be both in stimulation ROI and measurement ROI, we have the problem.

**Comment by WeiliangChenOIST** [comment]
From my test Gmsh 4 drops the extra tags inserted to gmsh 2 formated files, so it is unlikely that the feature is supported in the new version. In this case, we can acknowledge it as a restriction/limitation for the gmsh tagging implementation and mention it in the instruction.

**Comment by castigli** [comment]
Ok, I will add a comment in the documentation. Apart from that, the rest is ok?

**Comment by WeiliangChenOIST** [comment]
> Ok, I will add a comment in the documentation. Apart from that, the rest is ok?

has the other comments in this PR being addressed already?

**Comment by castigli** [comment]
> > Ok, I will add a comment in the documentation. Apart from that, the rest is ok?
> 
> has the other comments in this PR being addressed already?

I believe so, now previous tags are retained, but you can test it again to be sure it behaves as expected.

**Comment by WeiliangChenOIST** [comment]
> > > Ok, I will add a comment in the documentation. Apart from that, the rest is ok?
> > 
> > 
> > has the other comments in this PR being addressed already?
> 
> I believe so, now previous tags are retained, but you can test it again to be sure it behaves as expected.

actually I meant the version number and the usage explanation in the unresolved conversations above.

File: polyhedronROI/gmsh/tag_mesh_entities.py [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+    return spatial_index
+
+def add_tet_ROIs(mesh, import_scale, spatial_index, boundary_mesh_files, roi_labels):
+    """
+    Identify the tetrahedral elements in each ROI label, using a list of boundary mesh files.
+    Create each ROI in the gmsh.model.mesh object.
+
+    Parameters:
+        * mesh: a gmsh.model.mesh object
+        * import_scale: The scale used when importing the mesh data using steps.utilities.meshio
```
[the remark this answers is not in the document]
Comment (castigli) [reply polyhedronROI/gmsh/tag_mesh_entities.py:70]: > > > > Ok, I will add a comment in the documentation. Apart from that, the rest is ok?
> > > 
> > > 
> > > has the other comments in this PR being addressed already?
> > 
> > 
> > I believe so, now previous tags are retained, but you can test it again to be sure it behaves as expected.
> 
> actually I meant the version number and the usage explanation in the unresolved conversations above.

Ah! No, I will fix those. Also, I found a bug in the tagging, I will updated you when it is fixed and ready to be reviewed.

File: polyhedronROI/__init__.py (hunk 1 above)
Comment (castigli) [reply polyhedronROI/__init__.py:1]: done

File: example/gmsh/tests.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
+import polyhedronROI.gmsh as pRg
+import sys
+
+if __name__ == "__main__":
+    if len(sys.argv) < 2:
+        print("Usage: " + sys.argv[0] + " <cube|cube-overlap|spine> [--interactive]")
```
[the remark this answers is not in the document]
Comment (castigli) [reply example/gmsh/tests.py:7]: added the missing shebang, unfortunately in the usage error we cannot know if the script was invoked directly or with the python executable

File: polyhedronROI/gmsh/tag_mesh_entities.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (castigli) [reply polyhedronROI/gmsh/tag_mesh_entities.py:69]: removed scale

**Comment by castigli** [comment]
I fixed a couple of bugs and address the previous comments.
Before changing the documentation, I wanted to clarify what is the behavior that we want to keep: at the moment you can tag elements with multiple ROI, but overlapping tets are duplicated.
Is this ok, or do we want the default behavior for the gmsh version to be that you cannot have overlapping ROIs?

**Comment by WeiliangChenOIST** [comment]
> at the moment you can tag elements with multiple ROI, but overlapping tets are duplicated.

This solution doesn't seem to work very well with current gmsh as too much time is needed when loading such mesh due to the duplicated element issue. 

> default behavior for the gmsh version to be that you cannot have overlapping ROIs

I would set this as the default choice, i.e. an element can only be in a single ROI, and send an error if we find any element is in multiple ROIs. In this case, the user may have the choice of either updating the ROI definitions, or turning on a flag in the function to enable the duplicate mode.

**Comment by castigli** [comment]
Ok, now by default if a tet duplication is detected an error is raised (there is a flag  to allow duplication).
I also update the readme.

**Review by WeiliangChenOIST — APPROVED** [review]
LGTM

**Pull request merged**